### PR TITLE
[hab] Install same version of helper packages as hab if not present.

### LINKED ISSUE #1409: Ability to pin habitat versions
### Problem

While #1292 ensured that the hab studio image version matches that of the `hab` binary being used to launch it, there are still several cases where versions can get mismatched. For example:

```
$ ./hab --version
hab 0.10.2/20160930234832

$ ./hab studio enter

[1][default:/src:0]# hab --version
hab 0.10.2/20160930230245

[2][default:/src:0]# hab pkg install core/hab
» Installing core/hab
↓ Downloading core/hab/0.11.0/20161013211955
    2.33 MB / 2.33 MB - [====...] 100.00 % 8.41 MB/s
↓ Downloading core-20160810182414 public origin key
    75 B / 75 B | [====...] 100.00 % 2.02 MB/s
☑ Cached core-20160810182414 public origin key
✓ Installed core/hab/0.11.0/20161013211955
★ Install of core/hab/0.11.0/20161013211955 complete with 1 new packages installed.
```

Essentially, any reference to `core/hab` or `core/hab-studio` will always cause the latest release version to be pulled. This makes it difficult to stage the upgrade process in a build pipeline. One moment you are building against `0.10.2` and a few hours later you are building against `0.11.0` which is suddenly broken. This happened to us during that release cycle and it hit us hard -- in addition to the bad binary issue, we were affected by the changes to root/sudo as mentioned in #1408.
### Solution?

I propose that subshells launched from a given version of `hab` pin to that version. This could be done by exposing a new environment variable when hab is run and propagating that to any subshells as necessary. I'm not sure of all the implications of this as it would have to touch quite a few areas (hab, hab-sup, studio, export pkgs(?))

### LINKED ISSUE #1578: `hab start <myapp>` will start hab-sup 0.14 when hab is 0.13.1
```
[69][default:/src:0]# hab start bdangit/couchbase -C hab-plans/couchbase/
∵ Missing package for core/hab-sup
» Installing core/hab-sup
↓ Downloading core/hab-sup/0.14.0/20161212220457
    1.51 MB / 1.51 MB \ [=======================================================================================================] 100.00 % 172.11 KB/s
↓ Downloading core/busybox-static/1.24.2/20161209030307
    511.54 KB / 511.54 KB / [===================================================================================================] 100.00 % 377.60 KB/s
↓ Downloading core/bzip2/1.0.6/20161208225359
    57.99 KB / 57.99 KB | [=======================================================================================================] 100.00 % 3.39 MB/s
→ Using core/cacerts/2016.09.14/20161031044952
↓ Downloading core/gcc-libs/5.2.0/20161208223920
    1.74 MB / 1.74 MB | [=======================================================================================================] 100.00 % 833.26 KB/s
→ Using core/glibc/2.22/20160612063629
↓ Downloading core/libarchive/3.2.0/20161209032820
    542.68 KB / 542.68 KB / [=====================================================================================================] 100.00 % 1.20 MB/s
↓ Downloading core/libsodium/1.0.8/20161209224140
    181.30 KB / 181.30 KB \ [=====================================================================================================] 100.00 % 1.09 MB/s
→ Using core/linux-headers/4.3/20160612063537
↓ Downloading core/openssl/1.0.2j/20161208232101
    2.09 MB / 2.09 MB \ [=======================================================================================================] 100.00 % 552.07 KB/s
↓ Downloading core/xz/5.2.2/20161208230159
    236.43 KB / 236.43 KB / [===================================================================================================] 100.00 % 340.38 KB/s
↓ Downloading core/zeromq/4.1.4/20161210015641
    411.88 KB / 411.88 KB \ [===================================================================================================] 100.00 % 404.43 KB/s
↓ Downloading core/zlib/1.2.8/20161118033245
    71.56 KB / 71.56 KB / [=======================================================================================================] 100.00 % 2.40 MB/s
✓ Installed core/busybox-static/1.24.2/20161209030307
✓ Installed core/bzip2/1.0.6/20161208225359
✓ Installed core/gcc-libs/5.2.0/20161208223920
✓ Installed core/libarchive/3.2.0/20161209032820
✓ Installed core/libsodium/1.0.8/20161209224140
✓ Installed core/openssl/1.0.2j/20161208232101
✓ Installed core/xz/5.2.2/20161208230159
✓ Installed core/zeromq/4.1.4/20161210015641
✓ Installed core/zlib/1.2.8/20161118033245
✓ Installed core/hab-sup/0.14.0/20161212220457
★ Install of core/hab-sup/0.14.0/20161212220457 complete with 10 new packages installed.
hab-sup(MN): Starting bdangit/couchbase
hab-sup(MR): Butterfly Member ID 847c3cf293a34662a9a07e9728508f3f
couchbase.default(SR): Process will run as user=hab, group=hab
hab-sup(MR): Starting butterfly on 0.0.0.0:9638
hab-sup(MR): Starting http-gateway on 0.0.0.0:9631
hab-sup(SC): Updated config 0e5751c026e543b2e8ab2eb06099daa1d1e5df47778f7787faab45cdf12fe3a8
hab-sup(SC): Updated default.d-capi.ini 95ebaafa4e56fbed8d2c4b606b1bca8b81654def7e81fe82b7a58879fb319cab
hab-sup(SC): Updated default.d-geocouch.ini 71f4ca97a1ae14794533832e978ae30bc57f8d8ac1864f2c8bf66299861b75d2
hab-sup(SC): Updated default.ini ae22a779ed73b588e42b7b8b0ffb2e11036d2e27c73e4aefea8e2bfaca429194
hab-sup(SC): Updated local.ini f2a9ef197dee45ecaecf747be1e74de70c6ab75cfbbe7ed195add6b69e943c7b
hab-sup(SC): Updated pluggable-ui-fts.json b0ec1b65cf1de4d8888efefc5b84d66c46ce5796aaf37d616841473d162b122e
hab-sup(SC): Updated pluggable-ui-query.json 3ffe403b6e21519253ca1cd696494052ffd949d2a1b55aa2850292298b2d311d
hab-sup(SC): Updated static_config c8070490bd221a8ceb184af5628e4b471ce40a1f7da18ec05eb359a6099b9e47

[73][default:/src:1]# hab -V
hab 0.13.1/20161114235527

```

not expecting for hab-sup to be newer than the studio version.

### PR DESCRIPTION
**Pull request by fnichol**
This feature aims to give a user a more intuitive initial experience
when using a particular version of `hab`. This behavior currently only
applies to Linux targets as Windows and Mac have alternative or disabled
implementations. The following `hab` subcommands are affected:

* `hab pkg build` (delegates to `core/hab-studio` package)
* `hab studio` (delegates to `core/hab-studio` package)
* `hab sup` (delegates to `core/hab-sup` package)
* `hab start` (delegates to `core/hab-sup` package)
* `hab config apply` (delegates to `core/hab-butterfly` package)
* `hab file upload` (delegates to `core/hab-butterfly` package)
* `hab apply` (delegates to `core/hab-butterfly` package)

All of the above commands will install an initial version of their
helper package on first use. The previous behavior was to install the
latest version and release of the helper package. For example, running
`hab studio enter` would install the equivalent of first running `hab
install core/hab-studio`. Note that if the version of `hab` was `1.1`
and the latest version of `hab-studio` was `2.2`, then installing
`core/hab-studio` would yield `core/hab-studio/2.2/<RELEASE>` installed
and executed.

This change uses the `VERSION` coordinate of `hab`'s package identifier
when performing the initial installation. For example, given the
following version of `hab`:

    > hab --version
    hab 1.2.3/20170127132343

it would attempt to install the latest release of `core/hab-studio/1.2.3`
when running `hab pkg build`.

Favoring Later Versions On Disk
-------------------------------

There is an exception to the behavior described above. If, as a user on
a system (or some automated process), I choose to install a newer
version of the Supervisor (let's say `core/hab-sup/5.6.7`), then running
`hab sup start` would find this newer installed package and use that
rather than its `1.2.3` version (which may or may not be installed on
disk).

The feature is written in this way to prevent an unplanned pinning back
of the Supervisor during upgrades. In other words, if `hab` only ever
ran the same version of the Supervisor, the Studio, etc., a user would
have to remember to always update the version (and possibly symlink or
binlink) of `hab` in order to run the newer software.

The tl;dr here is:

* `hab` will only install the same version of `hab-sup`, `hab-studio`,
  `hab-butterfly` if a package is not already installed
* `hab` will not run a version of the above packages that is older than
  itself
* `hab` will use and run a version of the above packages if it is newer
  than itself, but only if previously installed

Closes #1409
Closes #1578

Signed-off-by: Fletcher Nichol <fnichol@nichol.ca>

### CODE REVIEW

**Comment by thesentinels** [comment]
Thanks for the pull request! Here is what will happen next:
 1. Your PR will be reviewed by the maintainers
 2. If everything looks good, one of them will `approve` it, and your PR will be merged.

Thank you for contributing!

**Comment by fnichol** [comment]
gif-keyboard-8017775986037747040

File: components/core/src/package/install.rs [hunk 1]
```diff
@@ -50,6 +50,22 @@ impl PackageInstall {
...
+    /// a Result of a `PackageIdent` if one exists.
+    ///
+    /// An optional `fs_root` path may be provided to search for a package that is mounted on a
+    /// filesystem not currently rooted at `/`.
+    pub fn load_at_least(ident: &PackageIdent,
+                         fs_root_path: Option<&Path>)
+                         -> Result<PackageInstall> {
+        let package_install = try!(Self::resolve_package_install_min(ident, fs_root_path));
+        let package_target = try!(package_install.target());
+        match package_target.validate() {
```
Comment (reset) [inline components/core/src/package/install.rs]: Since you're not using the inner type of a successful result of this call, it's probably better represented as

```
try!(package_target.validate());
Ok(package_install)
```

File: components/core/src/package/install.rs [hunk 2]
```diff
@@ -111,6 +127,57 @@ impl PackageInstall {
         }
     }
 
+    /// Find an installed package that is at minimum the version of the given ident.
+    fn resolve_package_install_min(ident: &PackageIdent,
+                                   fs_root_path: Option<&Path>)
```
Comment (reset) [inline components/core/src/package/install.rs]: We should try and get in the habit of using `AsRef<Path>` when we need `&Path`. There's a number of types in the standard library that implement that and it makes for a more usable UI. It's even a pattern in the Path module itself in the standard library ;)

**Comment by reset** [comment]
I left a few comments but nothing is a necessary change, just some rust-isms!

@thesentinels approve

**Comment by thesentinels** [comment]
:metal: I am testing your branch against master before merging it. We do this to ensure that the master branch is never failing tests.

**Comment by thesentinels** [comment]
:neckbeard: Travis CI has [started testing this PR](https://travis-ci.org/habitat-sh/habitat/builds/197515846).

**Comment by thesentinels** [comment]
:neckbeard: Travis CI has [started testing this PR](https://travis-ci.org/habitat-sh/habitat/builds/197515846).

**Comment by thesentinels** [comment]
:sparkling_heart: Travis CI [reports this PR passed](https://travis-ci.org/habitat-sh/habitat/builds/197515846).

It always makes me feel nice when humans approve of one anothers work. I'm merging this PR now.

I just want you and the contributor to answer me one question:

gif-keyboard-3280869874741411265

**Comment by fnichol** [comment]
Agreed with the review comments and felt a little dirty about copy pasta.
